sam/assembly.go

```diff
@@ -0,0 +1,219 @@
+package sam
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/numbers/logspace"
+	"log"
+	"math"
+)
+
+type DiploidBase byte
+
+const (
+	AA DiploidBase = 0
+	AC DiploidBase = 1
+	AG DiploidBase = 2
+	AT DiploidBase = 3
+	CC DiploidBase = 4
+	CG DiploidBase = 5
+	CT DiploidBase = 6
+	GG DiploidBase = 7
+	GT DiploidBase = 8
+	TT DiploidBase = 9
+	NN DiploidBase = 10
+)
+
+// diploidBaseString formats a DiploidBase struct as a string for debugging.
+func diploidBaseString(base DiploidBase) string {
+	switch base {
+	case AA:
+		return "AA"
+	case AC:
+		return "AC"
+	case AG:
+		return "AG"
+	case AT:
+		return "AT"
+	case CC:
+		return "CC"
+	case CG:
+		return "CG"
+	case CT:
+		return "CT"
+	case GG:
+		return "GG"
+	case GT:
+		return "GT"
+	case TT:
+		return "TT"
+	case NN:
+		return "NN"
+	}
+	log.Fatalf("Unrecognized DiploidBase: %v.", base)
+	return ""
+}
+
+//DiploidBaseCallFromPile determines which of 10 DiploidBase genotypes
```

Comment:
it would be helpful for the documentation to define where to get or how to derive the caches.